{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "from read_data import *\n",
    "from hyper import *\n",
    "from alignment import *\n",
    "from datasets.preprecossing import *\n",
    "import rpy2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'3.5.13'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rpy2.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "None\n"
     ]
    }
   ],
   "source": [
    "from rpy2.robjects.packages import importr\n",
    "import rpy2.robjects as robjects\n",
    "\n",
    "Spatialpc  = importr('SpatialPCA')\n",
    "ggplot2 = importr('ggplot2')\n",
    "ggplot2\n",
    "a = robjects.r('a<-c(1,2,3)')\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Use SCTransform function in Seurat to normalize data. \n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Calculating cell attributes from input UMI matrix: log_umi\n",
      "\n",
      "R[write to console]: Variance stabilizing transformation of count matrix of size 14262 by 4226\n",
      "\n",
      "R[write to console]: Model formula is y ~ log_umi\n",
      "\n",
      "R[write to console]: Get Negative Binomial regression parameters per gene\n",
      "\n",
      "R[write to console]: Using 2000 genes, 4226 cells\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Found 64 outliers - those will be ignored in fitting/regularization step\n",
      "\n",
      "\n",
      "R[write to console]: Second step: Get residuals using fitted parameters for 14262 genes\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Computing corrected count matrix for 14262 genes\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  |======================================================================| 100%\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Calculating gene attributes\n",
      "\n",
      "R[write to console]: Wall clock passed: Time difference of 48.41213 secs\n",
      "\n",
      "R[write to console]: Determine variable features\n",
      "\n",
      "R[write to console]: Place corrected count matrix in counts slot\n",
      "\n",
      "R[write to console]: Centering data matrix\n",
      "\n",
      "  |                                                                            \n",
      "  |                                                                      |   0%\n",
      "  |                                                                            \n",
      "  |====                                                                  |   5%\n",
      "  |                                                                            \n",
      "  |=======                                                               |  10%\n",
      "  |                                                                            \n",
      "  |==========                                                            |  15%\n",
      "  |                                                                            \n",
      "  |==============                                                        |  20%\n",
      "  |                                                                            \n",
      "  |==================                                                    |  25%\n",
      "  |                                                                            \n",
      "  |=====================                                                 |  30%\n",
      "  |                                                                            \n",
      "  |========================                                              |  35%\n",
      "  |                                                                            \n",
      "  |============================                                          |  40%\n",
      "  |                                                                            \n",
      "  |================================                                      |  45%\n",
      "  |                                                                            \n",
      "  |===================================                                   |  50%\n",
      "  |                                                                            \n",
      "  |======================================                                |  55%\n",
      "  |                                                                            \n",
      "  |==========================================                            |  60%\n",
      "  |                                                                            \n",
      "  |==============================================                        |  65%\n",
      "  |                                                                            \n",
      "  |=================================================                     |  70%\n",
      "  |                                                                            \n",
      "  |====================================================                  |  75%\n",
      "  |                                                                            \n",
      "  |========================================================              |  80%\n",
      "  |                                                                            \n",
      "  |============================================================          |  85%\n",
      "  |                                                                            \n",
      "  |===============================================================       |  90%\n",
      "  |                                                                            \n",
      "  |==================================================================    |  95%\n",
      "  |                                                                            \n",
      "  |======================================================================| 100%\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: Set default assay to SCT\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "## Use spark.test function in SPARK package to select spatially variable genes. \n",
      "## ===== SPARK INPUT INFORMATION ==== \n",
      "## number of total samples:  4226 \n",
      "## number of total features:  3974 \n",
      "## number of adjusted covariates:  0 \n",
      "# fitting normalized count-based spatial model under the null hypothesis ... \n",
      "## testing Gaussian kernel: 1...\n",
      "## testing Periodic kernel: 1...\n",
      "## testing Gaussian kernel: 2...\n",
      "## testing Periodic kernel: 2...\n",
      "## testing Gaussian kernel: 3...\n",
      "## testing Periodic kernel: 3...\n",
      "## testing Gaussian kernel: 4...\n",
      "## testing Periodic kernel: 4...\n",
      "## testing Gaussian kernel: 5...\n",
      "## testing Periodic kernel: 5...\n",
      "## Identified  1288  spatial genes through spark.test function. \n",
      "The  number of significant spatial genes is less than the specified number of spatial genes. \n",
      "## Using  1288  significant spatially variable genes. \n",
      "## Selected kernel type is:  gaussian  \n",
      "## The bandwidth is:  0.0387322493431883  \n",
      "## Calculating kernel matrix\n",
      "## Finished calculating kernel matrix.\n",
      "[1] \"Eigen decomposition on kernel matrix!\"\n",
      "[1] \"Using all eigenvectors and eigenvalues in the Kernel matrix!\"\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<rpy2.robjects.methods.RS4 object at 0x7fdb504eb3c0> [RTYPES.S4SXP]\n",
       "R classes: ('SpatialPCA',)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "robjects.r(\"load(paste0('./datas/pca/DLPFC/LIBD_sample1.RData'))\")\n",
    "robjects.r(\"xy_coords=as.matrix(xy_coords)\")\n",
    "robjects.r(\"rownames(xy_coords) = colnames(count_sub)\");\n",
    "robjects.r('LIBD = CreateSpatialPCAObject(counts=count_sub, location=xy_coords, project = \"SpatialPCA\",gene.type=\"spatial\",sparkversion=\"spark\",numCores_spark=5,gene.number=3000, customGenelist=NULL,min.loctions = 20, min.features=20)')\n",
    "robjects.r('LIBD = SpatialPCA_buildKernel(LIBD, kerneltype=\"gaussian\", bandwidthtype=\"SJ\",bandwidth.set.by.user=NULL)')\n",
    "robjects.r('LIBD = SpatialPCA_EstimateLoading(LIBD,fast=FALSE,SpatialPCnum=20) ')\n",
    "robjects.r('LIBD=SpatialPCA_SpatialPCs(LIBD, fast=FALSE)')\n",
    "# robjects.r('LIBD=SpatialPCA_SpatialPCs(LIBD, fast=FALSE)')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "robjects.r('write.csv(as.data.frame(LIBD@SpatialPCs), file = \"{}\")'.format(\"test.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 4220 × 9948\n",
       "    obs: 'in_tissue', 'array_row', 'array_col', 'imagerow', 'imagecol', 'sum_umi', 'sum_gene', 'subject', 'position', 'replicate', 'discard', 'cell_count', 'layer_guess', 'layer_guess_reordered', 'layer_guess_reordered_short', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'n_counts'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'n_counts'\n",
       "    uns: 'layer_guess_reordered_colors'\n",
       "    obsm: 'spatial'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datas = sc.read_h5ad(\"./datas/nature/151507_preprocessed.h5\")\n",
    "datas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "      <th>15</th>\n",
       "      <th>16</th>\n",
       "      <th>17</th>\n",
       "      <th>18</th>\n",
       "      <th>19</th>\n",
       "      <th>20</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACAACGAATAGTTC-1</th>\n",
       "      <td>1.839585</td>\n",
       "      <td>-0.056076</td>\n",
       "      <td>0.119012</td>\n",
       "      <td>2.259882</td>\n",
       "      <td>-0.725110</td>\n",
       "      <td>4.093265</td>\n",
       "      <td>1.573622</td>\n",
       "      <td>-1.778798</td>\n",
       "      <td>-0.037971</td>\n",
       "      <td>-2.481191</td>\n",
       "      <td>-0.346664</td>\n",
       "      <td>0.488325</td>\n",
       "      <td>-0.090961</td>\n",
       "      <td>0.254027</td>\n",
       "      <td>0.170660</td>\n",
       "      <td>-0.437685</td>\n",
       "      <td>-1.593910</td>\n",
       "      <td>1.668237</td>\n",
       "      <td>-0.665730</td>\n",
       "      <td>0.467312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAAGTATCTCCCA-1</th>\n",
       "      <td>-4.201764</td>\n",
       "      <td>-0.170601</td>\n",
       "      <td>-0.819079</td>\n",
       "      <td>-0.686475</td>\n",
       "      <td>1.279149</td>\n",
       "      <td>-1.377460</td>\n",
       "      <td>1.111393</td>\n",
       "      <td>0.318332</td>\n",
       "      <td>-0.016962</td>\n",
       "      <td>0.454175</td>\n",
       "      <td>0.764330</td>\n",
       "      <td>0.026984</td>\n",
       "      <td>0.718653</td>\n",
       "      <td>-0.332289</td>\n",
       "      <td>-0.268545</td>\n",
       "      <td>-0.011880</td>\n",
       "      <td>0.155666</td>\n",
       "      <td>0.085920</td>\n",
       "      <td>0.066624</td>\n",
       "      <td>0.266871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAATCTACTAGCA-1</th>\n",
       "      <td>8.301020</td>\n",
       "      <td>8.129599</td>\n",
       "      <td>-2.846293</td>\n",
       "      <td>-3.876941</td>\n",
       "      <td>-1.510169</td>\n",
       "      <td>0.406868</td>\n",
       "      <td>-0.511169</td>\n",
       "      <td>-0.609171</td>\n",
       "      <td>1.509160</td>\n",
       "      <td>1.234912</td>\n",
       "      <td>2.693397</td>\n",
       "      <td>-0.351943</td>\n",
       "      <td>0.188804</td>\n",
       "      <td>-0.748378</td>\n",
       "      <td>-0.216233</td>\n",
       "      <td>0.596782</td>\n",
       "      <td>0.166709</td>\n",
       "      <td>0.001538</td>\n",
       "      <td>-0.543832</td>\n",
       "      <td>1.452704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACACCAATAACTGC-1</th>\n",
       "      <td>15.506694</td>\n",
       "      <td>-9.032607</td>\n",
       "      <td>5.197555</td>\n",
       "      <td>-4.409749</td>\n",
       "      <td>2.281063</td>\n",
       "      <td>-1.550231</td>\n",
       "      <td>-0.261578</td>\n",
       "      <td>1.772977</td>\n",
       "      <td>0.224246</td>\n",
       "      <td>-0.274725</td>\n",
       "      <td>-0.701496</td>\n",
       "      <td>-1.173895</td>\n",
       "      <td>1.063487</td>\n",
       "      <td>0.523630</td>\n",
       "      <td>0.463621</td>\n",
       "      <td>0.154000</td>\n",
       "      <td>-1.093090</td>\n",
       "      <td>-0.941864</td>\n",
       "      <td>1.582432</td>\n",
       "      <td>0.393645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACAGCTTTCAGAAG-1</th>\n",
       "      <td>-2.607821</td>\n",
       "      <td>-1.672040</td>\n",
       "      <td>-0.047325</td>\n",
       "      <td>-0.060443</td>\n",
       "      <td>-3.080712</td>\n",
       "      <td>1.288785</td>\n",
       "      <td>1.136652</td>\n",
       "      <td>2.336629</td>\n",
       "      <td>-2.891868</td>\n",
       "      <td>2.962956</td>\n",
       "      <td>0.465117</td>\n",
       "      <td>-0.953662</td>\n",
       "      <td>-0.387020</td>\n",
       "      <td>0.941422</td>\n",
       "      <td>1.582869</td>\n",
       "      <td>0.892871</td>\n",
       "      <td>0.777105</td>\n",
       "      <td>0.299855</td>\n",
       "      <td>0.632999</td>\n",
       "      <td>-0.688456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTGTGTGTCAAGA-1</th>\n",
       "      <td>-3.570083</td>\n",
       "      <td>2.295234</td>\n",
       "      <td>0.850403</td>\n",
       "      <td>-1.855247</td>\n",
       "      <td>1.060709</td>\n",
       "      <td>-0.193066</td>\n",
       "      <td>-0.799369</td>\n",
       "      <td>-0.620065</td>\n",
       "      <td>0.082121</td>\n",
       "      <td>-0.674397</td>\n",
       "      <td>-0.035970</td>\n",
       "      <td>0.044762</td>\n",
       "      <td>-0.630693</td>\n",
       "      <td>-0.392696</td>\n",
       "      <td>0.485399</td>\n",
       "      <td>-0.082156</td>\n",
       "      <td>-0.230145</td>\n",
       "      <td>0.279488</td>\n",
       "      <td>0.728626</td>\n",
       "      <td>-0.108202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTTCACATCCAGG-1</th>\n",
       "      <td>0.154293</td>\n",
       "      <td>-3.614489</td>\n",
       "      <td>1.909944</td>\n",
       "      <td>-0.325207</td>\n",
       "      <td>-2.360692</td>\n",
       "      <td>-0.082183</td>\n",
       "      <td>-0.339690</td>\n",
       "      <td>-1.573506</td>\n",
       "      <td>-1.503427</td>\n",
       "      <td>1.239652</td>\n",
       "      <td>-0.406086</td>\n",
       "      <td>0.986256</td>\n",
       "      <td>-1.195355</td>\n",
       "      <td>-0.468261</td>\n",
       "      <td>0.300783</td>\n",
       "      <td>-0.376656</td>\n",
       "      <td>0.185939</td>\n",
       "      <td>-0.003332</td>\n",
       "      <td>0.433184</td>\n",
       "      <td>-0.166631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTTCATTAGTCTA-1</th>\n",
       "      <td>10.263456</td>\n",
       "      <td>-7.831578</td>\n",
       "      <td>1.197190</td>\n",
       "      <td>2.795934</td>\n",
       "      <td>2.954854</td>\n",
       "      <td>2.208829</td>\n",
       "      <td>-0.287341</td>\n",
       "      <td>-1.218083</td>\n",
       "      <td>-2.846245</td>\n",
       "      <td>-0.242626</td>\n",
       "      <td>1.136412</td>\n",
       "      <td>2.561909</td>\n",
       "      <td>2.410297</td>\n",
       "      <td>-2.321355</td>\n",
       "      <td>-1.923278</td>\n",
       "      <td>1.137040</td>\n",
       "      <td>1.663151</td>\n",
       "      <td>1.269623</td>\n",
       "      <td>-1.126727</td>\n",
       "      <td>1.399827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTTCCATACAACT-1</th>\n",
       "      <td>-1.416969</td>\n",
       "      <td>-2.258090</td>\n",
       "      <td>-0.229993</td>\n",
       "      <td>-0.779353</td>\n",
       "      <td>-1.631411</td>\n",
       "      <td>0.862991</td>\n",
       "      <td>-1.129650</td>\n",
       "      <td>0.348168</td>\n",
       "      <td>0.290358</td>\n",
       "      <td>0.698683</td>\n",
       "      <td>-0.660364</td>\n",
       "      <td>-0.642317</td>\n",
       "      <td>0.723156</td>\n",
       "      <td>-0.166788</td>\n",
       "      <td>1.208592</td>\n",
       "      <td>0.245354</td>\n",
       "      <td>0.020271</td>\n",
       "      <td>-0.538028</td>\n",
       "      <td>-0.094238</td>\n",
       "      <td>-0.850288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTGTTTGTGTAAATTC-1</th>\n",
       "      <td>4.788981</td>\n",
       "      <td>6.892154</td>\n",
       "      <td>-1.682012</td>\n",
       "      <td>-2.019620</td>\n",
       "      <td>-0.839887</td>\n",
       "      <td>0.323372</td>\n",
       "      <td>-2.442276</td>\n",
       "      <td>0.490276</td>\n",
       "      <td>0.125759</td>\n",
       "      <td>0.634976</td>\n",
       "      <td>1.569091</td>\n",
       "      <td>0.838489</td>\n",
       "      <td>-0.929575</td>\n",
       "      <td>-0.179612</td>\n",
       "      <td>-0.586279</td>\n",
       "      <td>-0.514470</td>\n",
       "      <td>-0.663970</td>\n",
       "      <td>0.690898</td>\n",
       "      <td>0.342729</td>\n",
       "      <td>0.167473</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4226 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Unnamed: 0                 1         2         3         4         5   \\\n",
       "AAACAACGAATAGTTC-1   1.839585 -0.056076  0.119012  2.259882 -0.725110   \n",
       "AAACAAGTATCTCCCA-1  -4.201764 -0.170601 -0.819079 -0.686475  1.279149   \n",
       "AAACAATCTACTAGCA-1   8.301020  8.129599 -2.846293 -3.876941 -1.510169   \n",
       "AAACACCAATAACTGC-1  15.506694 -9.032607  5.197555 -4.409749  2.281063   \n",
       "AAACAGCTTTCAGAAG-1  -2.607821 -1.672040 -0.047325 -0.060443 -3.080712   \n",
       "...                       ...       ...       ...       ...       ...   \n",
       "TTGTTGTGTGTCAAGA-1  -3.570083  2.295234  0.850403 -1.855247  1.060709   \n",
       "TTGTTTCACATCCAGG-1   0.154293 -3.614489  1.909944 -0.325207 -2.360692   \n",
       "TTGTTTCATTAGTCTA-1  10.263456 -7.831578  1.197190  2.795934  2.954854   \n",
       "TTGTTTCCATACAACT-1  -1.416969 -2.258090 -0.229993 -0.779353 -1.631411   \n",
       "TTGTTTGTGTAAATTC-1   4.788981  6.892154 -1.682012 -2.019620 -0.839887   \n",
       "\n",
       "Unnamed: 0                6         7         8         9         10  \\\n",
       "AAACAACGAATAGTTC-1  4.093265  1.573622 -1.778798 -0.037971 -2.481191   \n",
       "AAACAAGTATCTCCCA-1 -1.377460  1.111393  0.318332 -0.016962  0.454175   \n",
       "AAACAATCTACTAGCA-1  0.406868 -0.511169 -0.609171  1.509160  1.234912   \n",
       "AAACACCAATAACTGC-1 -1.550231 -0.261578  1.772977  0.224246 -0.274725   \n",
       "AAACAGCTTTCAGAAG-1  1.288785  1.136652  2.336629 -2.891868  2.962956   \n",
       "...                      ...       ...       ...       ...       ...   \n",
       "TTGTTGTGTGTCAAGA-1 -0.193066 -0.799369 -0.620065  0.082121 -0.674397   \n",
       "TTGTTTCACATCCAGG-1 -0.082183 -0.339690 -1.573506 -1.503427  1.239652   \n",
       "TTGTTTCATTAGTCTA-1  2.208829 -0.287341 -1.218083 -2.846245 -0.242626   \n",
       "TTGTTTCCATACAACT-1  0.862991 -1.129650  0.348168  0.290358  0.698683   \n",
       "TTGTTTGTGTAAATTC-1  0.323372 -2.442276  0.490276  0.125759  0.634976   \n",
       "\n",
       "Unnamed: 0                11        12        13        14        15  \\\n",
       "AAACAACGAATAGTTC-1 -0.346664  0.488325 -0.090961  0.254027  0.170660   \n",
       "AAACAAGTATCTCCCA-1  0.764330  0.026984  0.718653 -0.332289 -0.268545   \n",
       "AAACAATCTACTAGCA-1  2.693397 -0.351943  0.188804 -0.748378 -0.216233   \n",
       "AAACACCAATAACTGC-1 -0.701496 -1.173895  1.063487  0.523630  0.463621   \n",
       "AAACAGCTTTCAGAAG-1  0.465117 -0.953662 -0.387020  0.941422  1.582869   \n",
       "...                      ...       ...       ...       ...       ...   \n",
       "TTGTTGTGTGTCAAGA-1 -0.035970  0.044762 -0.630693 -0.392696  0.485399   \n",
       "TTGTTTCACATCCAGG-1 -0.406086  0.986256 -1.195355 -0.468261  0.300783   \n",
       "TTGTTTCATTAGTCTA-1  1.136412  2.561909  2.410297 -2.321355 -1.923278   \n",
       "TTGTTTCCATACAACT-1 -0.660364 -0.642317  0.723156 -0.166788  1.208592   \n",
       "TTGTTTGTGTAAATTC-1  1.569091  0.838489 -0.929575 -0.179612 -0.586279   \n",
       "\n",
       "Unnamed: 0                16        17        18        19        20  \n",
       "AAACAACGAATAGTTC-1 -0.437685 -1.593910  1.668237 -0.665730  0.467312  \n",
       "AAACAAGTATCTCCCA-1 -0.011880  0.155666  0.085920  0.066624  0.266871  \n",
       "AAACAATCTACTAGCA-1  0.596782  0.166709  0.001538 -0.543832  1.452704  \n",
       "AAACACCAATAACTGC-1  0.154000 -1.093090 -0.941864  1.582432  0.393645  \n",
       "AAACAGCTTTCAGAAG-1  0.892871  0.777105  0.299855  0.632999 -0.688456  \n",
       "...                      ...       ...       ...       ...       ...  \n",
       "TTGTTGTGTGTCAAGA-1 -0.082156 -0.230145  0.279488  0.728626 -0.108202  \n",
       "TTGTTTCACATCCAGG-1 -0.376656  0.185939 -0.003332  0.433184 -0.166631  \n",
       "TTGTTTCATTAGTCTA-1  1.137040  1.663151  1.269623 -1.126727  1.399827  \n",
       "TTGTTTCCATACAACT-1  0.245354  0.020271 -0.538028 -0.094238 -0.850288  \n",
       "TTGTTTGTGTAAATTC-1 -0.514470 -0.663970  0.690898  0.342729  0.167473  \n",
       "\n",
       "[4226 rows x 20 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datas = pd.read_csv(\"./my_list.csv\")\n",
    "datas=datas.set_index(datas.columns[0]).T\n",
    "datas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['in_tissue', 'array_row', 'array_col', 'imagerow', 'imagecol',\n",
       "       'sum_umi', 'sum_gene', 'subject', 'position', 'replicate', 'discard',\n",
       "       'cell_count', 'layer_guess', 'layer_guess_reordered',\n",
       "       'layer_guess_reordered_short', 'n_genes_by_counts',\n",
       "       'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts',\n",
       "       'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes',\n",
       "       'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes',\n",
       "       'n_counts'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datas.obs.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    4220.0\n",
       "mean        0.0\n",
       "std         0.0\n",
       "min         0.0\n",
       "25%         0.0\n",
       "50%         0.0\n",
       "75%         0.0\n",
       "max         0.0\n",
       "Name: position, dtype: float64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
